### ISSUE TITLE
Sniff include patterns never match content from STDIN

### ISSUE THREAD

**Issue by hacktivista** [squizlabs/PHP_CodeSniffer#3107]
**Describe the bug**

Some rules are being ignored if I try to inspect code from STDIN, but are correctly detected if I pass the filename as argument.

While using https://github.com/WPTT/WPThemeReview

**Code sample**
```php
<?php
function any_function(): void {
	// nothing yet
}
```

**To reproduce**
1. Create a file called `test.php` with the code sample above
2. Run `phpcs --report=csv --standard=WPThemeReview test.php`
3. Run `cat test.php | phpcs --report=csv --standard=WPThemeReview -`
4. Compare results
```
$ vendor/bin/phpcs --report=csv --standard=WPThemeReview test.php
File,Line,Column,Type,Message,Source,Severity,Fixable
"/home/me/proj/test.php",2,26,error,"void return type is not present in PHP version 7.0 or earlier",PHPCompatibility.FunctionDeclarations.NewReturnTypeDeclarations.voidFound,5,0
$ cat test.php | vendor/bin/phpcs --report=csv --standard=WPThemeReview
File,Line,Column,Type,Message,Source,Severity,Fixable
```

**Expected behavior**
It should display the same errors while reading from file or STDIN.

**Versions (please complete the following information):**
 - OS: Artix Linux (Arch without systemd)
 - PHP: 7.4.9
 - PHPCS: 3.5.6
 - Standard: WPThemeReview

**Comment by jrfnl** [comment]
@hacktivista I'm not familiar with `cat`, but IIRC, you need to tell PHPCS to use STDIN by using `-`.

As per the help which can be displayed using `--help`. Also see: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Usage

**Comment by jrfnl** [comment]
The only other thing I can think of is that `STDIN` is being ignored because of the `<include-pattern>*\.(php|inc)$</include-pattern>` setting in the WPThemeReview ruleset, which is applied to the sniffs which your error is coming from (`PHPCompatibility`).

That pattern is in place as WP Themes often have a large number of JS/CSS files as well and the checks from PHPCompatibility do not apply to those.

I would expect those kind of include patterns to be ignored though when the file is passed via STDIN.
@gsherwood Is there a way to have an `include-pattern` without breaking STDIN ? Is this a bug or was this implemented like this by design ?

**Comment by hacktivista** [comment]
I've tried with `-` and without it, its the same result

**Comment by jrfnl** [comment]
@hacktivista I've pulled a patch now to prevent `include-pattern`s being applied to STDIN. Works on my machines, but testing would be appreciated.

**Comment by jrfnl** [comment]
P.S.: Thanks for reporting this @hacktivista !

**Comment by gsherwood** [comment]
To tell PHPCS what the file path should be for a piece of content passed via STDIN, use the `--stdin-path` CLI argument:
```
cat test.php | vendor/bin/phpcs --report=csv --standard=WPThemeReview --stdin-path=/path/to/temp.php
```
Note that the path doesn't actually have to exist. PHPCS wont check it, but it will use it as the file name and path throughout the code.

Having this file path set let's PHPCS exclude files based on whatever rules you have provided, and allows sniffs that access the file path to work correctly.

**Comment by jrfnl** [comment]
@gsherwood Seeing your reply, should I adjust the PR to take `--stdin-path` into account if available ?

**Comment by gsherwood** [comment]
> @gsherwood Seeing your reply, should I adjust the PR to take `--stdin-path` into account if available ?

I don't know if any changes would be required. By the time that code runs, the file path should be set to the fake value (I think) so it might work without modification. I haven't had time to look.

**Comment by jrfnl** [comment]
I've just tested the patch in #3108 with the above in mind and IMO no additional changes are needed.

Tested as follows:
1. Created a file with the above mentioned code called `phpcs-3107.php`.
2. Checked out the PHPCS `master` branch.
3. Ran the following commands:
    1. `phpcs -ps ./phpcs-3107.php --standard=WPThemeReview`
        As expected it shows 1 error against the `PHPCompatibility.FunctionDeclarations.NewReturnTypeDeclarations` sniff.
    2. `cat ./phpcs-3107.php | phpcs -ps --standard=wpthemereview -`
        As reported above, no error is shown.
    3. `cat ./phpcs-3107.php | phpcs -ps --standard=wpthemereview --stdin-path=temp.php -`
        As expected based on the above info provided by @gsherwood, the 1 error is shown.
4. Checked out the `feature/3107-dont-check-stdin-against-ignore-include-paths` PR branch.
5. Ran all three commands again and this time the error is (correctly) shown in all three situations.

**Comment by jrfnl** [comment]
So to be clear: with the proposed patch applied, if the `--stdin-path=` argument is NOT provided, file name based `exclude` rules are not applied to `STDIN` and the code provided via `STDIN` will be scanned against ALL rules.

**Comment by gsherwood** [comment]
I've decided not to include this in 3.5.7, and instead postpone for 3.6 pending some more thought.

Include patterns for sniffs say "only run this sniff if the file path matches this pattern". When there is no file path to match, I'm not 100% sure that the rule should actually be applied. Exclude patterns should probably be ignored, but include patterns are different.

Either way, this is a BC break so it can't be in a bug fix release.

I think I'm leaning towards accepting the change and making it a rule that all checks based on file path will be ignored if the file path is not set. So no exclusions will take place and all sniffs will run. Internally, sniffs are allowed to check the file path and apply different logic, but it would be up to them to handle STDIN.

**Comment by gsherwood** [comment]
The change has now been merged in again, so properly closing now. The change will be in 3.6.0.
